src/main/java/org/openjdk/shenandoah/DataProvider.java

```diff
@@ -24,58 +24,97 @@
+    private final DataConnector connector;
 
-    private final int maxRegions;
-    private final long maxSize;
-    private final LongMonitor[] data;
-    private final StringMonitor[] matrix;
-    private final LongMonitor status;
-
-    public DataProvider(String id) throws Exception {
-        MonitoredHost host = MonitoredHost.getMonitoredHost(id);
-        MonitoredVm vm = host.getMonitoredVm(new VmIdentifier(id));
-        LongMonitor timestamp = (LongMonitor) vm.findByName("sun.gc.shenandoah.regions.timestamp");
-        LongMonitor max_regions_mon = (LongMonitor) vm.findByName("sun.gc.shenandoah.regions.max_regions");
+    private int maxRegions;
+    private long protocolVersion;
+    private long maxSize;
+    private LongMonitor[] data;
+    private LongMonitor status;
+
+
+
+    public DataProvider(String id) {
+        connector = new DataConnector(this::setMonitoredVm);
+        connector.start();
+    }
+
+    private void setMonitoredVm(MonitoredVm vm) {
+        LongMonitor max_regions_mon = getMonitor(vm,"sun.gc.shenandoah.regions.max_regions");
         maxRegions = (int) max_regions_mon.longValue();
-        LongMonitor max_size_mon = (LongMonitor) vm.findByName("sun.gc.shenandoah.regions.region_size");
+        //Reads in the version of the garbage collector
+        LongMonitor protocol_version_mon = getMonitor(vm, "sun.gc.shenandoah.regions.protocol_version");
+        if (protocol_version_mon == null) {
+            protocolVersion = ORIGINAL_VERSION;
+        } else {
+            protocolVersion = protocol_version_mon.longValue();
+        }
+        //System.out.println("The version of the Shenandoah is " + protocolVersion);
+        LongMonitor max_size_mon = getMonitor(vm,"sun.gc.shenandoah.regions.region_size");
         maxSize = max_size_mon.longValue();
-        status = (LongMonitor) vm.findByName("sun.gc.shenandoah.regions.status");
+        status = getMonitor(vm, "sun.gc.shenandoah.regions.status");
 
         data = new LongMonitor[maxRegions];
-        matrix = new StringMonitor[maxRegions];
         for (int i = 0; i < maxRegions; i++) {
-            LongMonitor mon = (LongMonitor) vm.findByName("sun.gc.shenandoah.regions.region." + i + ".data");
+            LongMonitor mon = getMonitor(vm,"sun.gc.shenandoah.regions.region." + i + ".data");
             if (mon != null) {
                 data[i] = mon;
             } else {
                 throw new IllegalStateException("Insufficient shared memory for all region counters. " +
                         "Try -XX:PerfDataMemorySize=512K or higher when running the monitored program.");
             }
+        }
+    }
 
-            StringMonitor mtrx = (StringMonitor) vm.findByName("sun.gc.shenandoah.regions.region." + i + ".matrix");
-            if (mtrx != null) {
-                matrix[i] = mtrx;
-            }
+    private static <T> T getMonitor(MonitoredVm vm, String key) {
+        try {
+            //noinspection unchecked
```

Comment:
What is this comment trying to say?

Reply:
This is an IDE directive to stop it from complaining about the unchecked cast. I can remove it.